storage/src/main/java/tech/pegasys/teku/storage/server/kvstore/dataaccess/VersionedHashDBSource.java

```diff
@@ -0,0 +1,319 @@
+/*
+ * Copyright Consensys Software Inc., 2025
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.teku.storage.server.kvstore.dataaccess;
+
+import static com.google.common.base.Preconditions.checkArgument;
+
+import com.google.common.primitives.Longs;
+import java.util.Optional;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.function.Function;
+import java.util.function.Supplier;
+import org.apache.commons.lang3.tuple.Pair;
+import org.apache.tuweni.bytes.Bytes;
+import org.apache.tuweni.bytes.Bytes32;
+import tech.pegasys.teku.infrastructure.unsigned.UInt64;
+import tech.pegasys.teku.spec.Spec;
+import tech.pegasys.teku.spec.SpecMilestone;
+import tech.pegasys.teku.spec.datastructures.blobs.versions.deneb.BlobSidecar;
+import tech.pegasys.teku.spec.datastructures.blobs.versions.fulu.DataColumnSidecar;
+import tech.pegasys.teku.spec.datastructures.util.DataColumnSlotAndIdentifier;
+import tech.pegasys.teku.spec.datastructures.util.SlotAndBlockRootAndBlobIndex;
+import tech.pegasys.teku.spec.logic.versions.deneb.types.VersionedHash;
+import tech.pegasys.teku.spec.schemas.SchemaDefinitionsDeneb;
+import tech.pegasys.teku.spec.schemas.SchemaDefinitionsFulu;
+import tech.pegasys.teku.storage.server.kvstore.serialization.UInt64Serializer;
+
+public class VersionedHashDBSource {
+  private final Bytes CANONICAL_BLOB_SIDECAR_TYPE = Bytes.of(0x00);
+  private final Bytes NON_CANONICAL_BLOB_SIDECAR_TYPE = Bytes.of(0x10);
+  private final Bytes CANONICAL_DATA_COLUMN_SIDECAR_TYPE = Bytes.of(0x20);
+  private final Bytes NON_CANONICAL_DATA_COLUMN_SIDECAR_TYPE = Bytes.of(0x30);
+
+  static final int TYPE_SIZE = 1;
+  static final int SLOT_SIZE = Long.BYTES;
+  static final int BLOCK_ROOT_SIZE = Bytes32.SIZE;
+  static final int BLOB_INDEX_SIZE = Long.BYTES;
+
+  static final int SLOT_OFFSET = TYPE_SIZE;
+  static final int BLOCK_ROOT_OFFSET = SLOT_OFFSET + SLOT_SIZE;
+  static final int BLOB_INDEX_OFFSET = BLOCK_ROOT_OFFSET + BLOCK_ROOT_SIZE;
+  static final int DATA_SIZE = BLOB_INDEX_OFFSET + BLOB_INDEX_SIZE;
+
+  private final Function<BlobSidecar, VersionedHash> blobSidecarToVersionedHash;
+  private final Function<Pair<DataColumnSidecar, UInt64>, VersionedHash>
+      dataColumnSidecarToVersionedHash;
+  private final AtomicBoolean storeSidecarHashes = new AtomicBoolean(false);
```

Comment:
shouldn't we control the activation of the blobSidecar's versioned hashes storage too? not sure, we want to activate this for blobs by default...

Reply:
It replaces `getBlobSidecars`, it's not in debug space https://github.com/ethereum/beacon-APIs/pull/546
So by spec we should support it by default from my understanding